gengo/src/lib.rs

```diff
@@ -69,16 +69,18 @@ pub struct Gengo<FS: for<'fs> FileSource<'fs>> {
 impl<FS: for<'fs> FileSource<'fs>> Gengo<FS> {
     /// Analyzes each file in the repository at the given revision.
     pub fn analyze(&self) -> Result<Analysis> {
+        let state = self.file_source.state()?;
         let entries: Vec<(_, _)> = self
             .file_source
             .entries()?
             .par_bridge()
-            .filter_map(|entry| {
-                let filepath = self.file_source.filepath(&entry).ok()?;
-                let contents = self.file_source.contents(&entry).ok()?;
-                let entry = self.analyze_blob(&filepath, contents)?;
+            .map_with(state, |state, entry| {
```

Comment:
It's interesting that it *looks like* the state is passed by reference, even though it's definitely not `Sync`.
But ignoring that, this most definitely fixes the cloning issue.

Now the last improvement would be to work on chunks so each threads sees more than one path of the *ordered* list of inputs (at least in case of an index iteration).

Reply:
The docs weren't 100% clear, but they say that `map_with` will clone the `INIT` (`state`) "as needed." Without diving into the source code I can't be sure, but I *interpreted* this as meaning that it will clone `state` per thread. I can't imagine this being reasonable unless rayon was already chunking up the iteration among threads, with each thread passing its state as a mutable reference to each iteration.

Reply:
That's sounds exactly like it should to have thread-local state. The performance also speaks for itself. Chunking should be easy and is doable by using [`par_chunks(size)`](https://docs.rs/rayon/latest/rayon/slice/trait.ParallelSlice.html#method.par_chunks) if you ever feel like getting the remaining 7% or more. Maybe rayon can do things better than the home-grown parallelization `gix` is using, so that would be interesting.